pkg/fake/iamapi.go

```diff
@@ -0,0 +1,150 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/iam/iamiface"
+)
+
+const ()
+
+// IAMAPIBehavior must be reset between tests otherwise tests will
+// pollute each other.
+type IAMAPIBehavior struct {
+	GetInstanceProfileBehavior            MockedFunction[iam.GetInstanceProfileInput, iam.GetInstanceProfileOutput]
+	CreateInstanceProfileBehavior         MockedFunction[iam.CreateInstanceProfileInput, iam.CreateInstanceProfileOutput]
+	DeleteInstanceProfileBehavior         MockedFunction[iam.DeleteInstanceProfileInput, iam.DeleteInstanceProfileOutput]
+	AddRoleToInstanceProfileBehavior      MockedFunction[iam.AddRoleToInstanceProfileInput, iam.AddRoleToInstanceProfileOutput]
+	RemoveRoleFromInstanceProfileBehavior MockedFunction[iam.RemoveRoleFromInstanceProfileInput, iam.RemoveRoleFromInstanceProfileOutput]
+}
+
+type IAMAPI struct {
+	sync.Mutex
+
+	iamiface.IAMAPI
+	IAMAPIBehavior
+
+	InstanceProfiles map[string]*iam.InstanceProfile
+}
+
+func NewIAMAPI() *IAMAPI {
+	return &IAMAPI{InstanceProfiles: map[string]*iam.InstanceProfile{}}
+}
+
+// Reset must be called between tests otherwise tests will pollute
+// each other.
+func (s *IAMAPI) Reset() {
+	s.GetInstanceProfileBehavior.Reset()
+	s.CreateInstanceProfileBehavior.Reset()
+	s.DeleteInstanceProfileBehavior.Reset()
+	s.AddRoleToInstanceProfileBehavior.Reset()
+	s.RemoveRoleFromInstanceProfileBehavior.Reset()
+	s.InstanceProfiles = map[string]*iam.InstanceProfile{}
+}
+
+func (s *IAMAPI) GetInstanceProfileWithContext(_ context.Context, input *iam.GetInstanceProfileInput, _ ...request.Option) (*iam.GetInstanceProfileOutput, error) {
+	return s.GetInstanceProfileBehavior.Invoke(input, func(*iam.GetInstanceProfileInput) (*iam.GetInstanceProfileOutput, error) {
+		s.Lock()
+		defer s.Unlock()
```

Comment:
Definitely good practice to add these locks/unlocks into our mock APIs. Not saying you need to do in this PR, but we should probably do it.